Gems/Atom/RHI/Code/Source/RHI/MultiDeviceTransientAttachmentPool.cpp

```diff
@@ -0,0 +1,387 @@
+            RemoveFromCache(descriptor.m_attachmentId);
+
+            // Ownership is managed by the cache.
+            RHI::Ptr<MultiDeviceImage> imagePtr = aznew MultiDeviceImage();
+            image = imagePtr.get();
+
+            imagePtr->Init(GetDeviceMask());
+
+            for (auto& [deviceIndex, deviceTransientAttachmentPool] : m_deviceTransientAttachmentPools)
+            {
+                imagePtr->m_deviceObjects[deviceIndex] = deviceTransientAttachmentPool->ActivateImage(descriptor);
+                if (imagePtr->GetDeviceImage(deviceIndex))
+                {
+                    imagePtr->SetDescriptor(imagePtr->GetDeviceImage(deviceIndex)->GetDescriptor());
+                }
+            }
+
+            imagePtr->SetName(descriptor.m_attachmentId);
+            m_cache.Insert(static_cast<uint64_t>(hash), AZStd::move(imagePtr));
+            if (!descriptor.m_attachmentId.IsEmpty())
+            {
+                m_reverseLookupHash.emplace(descriptor.m_attachmentId, hash);
+            }
+        }
+
+        return image;
+    }
+
+    MultiDeviceBuffer* MultiDeviceTransientAttachmentPool::ActivateBuffer(const TransientBufferDescriptor& descriptor)
+    {
+        MultiDeviceBuffer* buffer = nullptr;
+
+        HashValue64 hash = descriptor.GetHash();
+
+        // Check the cache for the buffer.
+        if (auto attachment = m_cache.Find(static_cast<uint64_t>(hash)))
+        {
+            buffer = static_cast<MultiDeviceBuffer*>(attachment);
+        }
+        // buffer is not in cache. Create a new one at the placed address, and add it to the cache.
+        else
+        {
+            // Remove any existing resource entries from the cache before adding a new one
+            RemoveFromCache(descriptor.m_attachmentId);
+
+            // Ownership is managed by the cache.
+            RHI::Ptr<MultiDeviceBuffer> bufferPtr = aznew MultiDeviceBuffer();
+            buffer = bufferPtr.get();
+
+            bufferPtr->Init(GetDeviceMask());
+
+            for (auto& [deviceIndex, deviceTransientAttachmentPool] : m_deviceTransientAttachmentPools)
+            {
+                bufferPtr->m_deviceObjects[deviceIndex] = deviceTransientAttachmentPool->ActivateBuffer(descriptor);
+                if (bufferPtr->GetDeviceBuffer(deviceIndex))
+                {
+                    bufferPtr->SetDescriptor(bufferPtr->GetDeviceBuffer(deviceIndex)->GetDescriptor());
+                }
+            }
+
+            bufferPtr->SetName(descriptor.m_attachmentId);
+            m_cache.Insert(static_cast<uint64_t>(hash), AZStd::move(bufferPtr));
+            if (!descriptor.m_attachmentId.IsEmpty())
+            {
+                m_reverseLookupHash.emplace(descriptor.m_attachmentId, hash);
+            }
+        }
+
+        return buffer;
+    }
+
+    void MultiDeviceTransientAttachmentPool::DeactivateBuffer(const AttachmentId& attachmentId)
+    {
+        RemoveFromCache(attachmentId);
```

Comment:
Are you sure you want to remove from cache? This method may get called multiple times a frame.